### PR TITLE
[FIX] Nonce Too Low for Approve Transaction

### PR DESCRIPTION
**Pull request by blackdevelopa**
**Development & PR Process**
1. Follow MetaMask [Mobile Coding Standards](https://docs.google.com/document/d/1VJLwTRsUw_5EDq_o8d6sSbXUAYENLurkRitYO45eY5o/edit?usp=sharing)
2. Add `release-xx` label to identify the PR slated for a upcoming release (will be used in release discussion)
3. Add `needs-dev-review` label when work is completed
4. Add `needs-qa` label when dev review is completed
5. Add `QA Passed` label when QA has signed off

**Description**
If CustomNonce is toggled to true in Advanced Settings and a nonce value is defined when approving a dapp, the same nonce value is reused when approving token allowance. This causes the transaction to fail and return `nonce too low` or `Internal JSON-RPC error`. We also do not have the option to increment this value when approving token.

**Screenshots/Recordings**
Before
https://recordit.co/tIxFpT7ohi

After
https://recordit.co/f3aEPhFrZp

_If applicable, add screenshots and/or recordings to visualize the before and after of your change_

**Issue**

Progresses https://github.com/MetaMask/metamask-mobile/issues/6017

**Checklist**

* [x] There is a related GitHub issue
* [ ] Tests are included if applicable
* [ ] Any added code is fully documented

### CODE REVIEW

File: app/components/Views/ApproveView/Approve/index.js [hunk 1]
```diff
@@ -350,8 +354,10 @@ class Approve extends PureComponent {
   };
 
   prepareTransaction = (transaction) => {
-    const { gasEstimateType } = this.props;
+    const { gasEstimateType, showCustomNonce } = this.props;
     const { legacyGasTransaction, eip1559GasTransaction } = this.state;
+    const { nonce } = transaction;
+    if (showCustomNonce && nonce) transaction.nonce = BNToHex(nonce + 1);
```
Comment (blackdevelopa) [inline app/components/Views/ApproveView/Approve/index.js:360]: This will be refactored

Comment (blackdevelopa) [reply app/components/Views/ApproveView/Approve/index.js]: Cleaned up

**Comment by jpuri** [comment]
I am thinking if this logic can be put in 1 place rather than repeating in 3 react components.

File: app/components/UI/TransactionReview/TransactionReviewInformation/index.js [hunk 2]
```diff
@@ -221,15 +222,10 @@ class TransactionReviewInformation extends PureComponent {
...
-
-  setNetworkNonce = async () => {
-    const { setNonce, setProposedNonce, transaction } = this.props;
-    const proposedNonce = await getNetworkNonce(transaction);
-    setNonce(proposedNonce);
-    setProposedNonce(proposedNonce);
+    const { showCustomNonce, setNonce, setProposedNonce, transaction } =
+      this.props;
+    showCustomNonce &&
+      (await NetworkNonce({ setNonce, setProposedNonce, transaction }));
```
[the remark this answers is not in the document]
Comment (blackdevelopa) [reply app/components/UI/TransactionReview/TransactionReviewInformation/index.js]: I think the performance difference isn't a lot, but happy to make this switch. Here https://github.com/MetaMask/metamask-mobile/pull/5718/commits/95f741c56e0da0fccd7708b35f62746e316d8d85

File: app/components/Views/ApproveView/Approve/index.js [hunk 3]
```diff
@@ -192,14 +209,17 @@ class Approve extends PureComponent {
...
+    const { setNonce, setProposedNonce, transaction } = this.props;
     if (!this.props?.transaction?.id) {
       this.props.toggleApproveModal(false);
       return null;
     }
     if (!this.props?.transaction?.gas) this.handleGetGasLimit();
 
     this.startPolling();
+    this.props.showCustomNonce &&
+      (await NetworkNonce({ setNonce, setProposedNonce, transaction }));
```
[the remark this answers is not in the document]
Comment (blackdevelopa) [reply app/components/Views/ApproveView/Approve/index.js]: Updated. https://github.com/MetaMask/metamask-mobile/pull/5718/commits/95f741c56e0da0fccd7708b35f62746e316d8d85

**Comment by blackdevelopa** [comment]
> I am thinking if this logic can be put in 1 place rather than repeating in 3 react components.

I think it is? No? [Here](https://github.com/MetaMask/metamask-mobile/blob/95f741c56e0da0fccd7708b35f62746e316d8d85/app/util/networks/networkNonce.ts).

File: app/components/Views/ApproveView/Approve/index.js [hunk 4]
```diff
@@ -192,15 +209,19 @@ class Approve extends PureComponent {
     this.setState({ pollToken });
   };
 
-  componentDidMount = () => {
+  componentDidMount = async () => {
```
Comment (digiwand) [inline app/components/Views/ApproveView/Approve/index.js:212]:
```suggestion
  componentDidMount = () => {
```

**Review by digiwand — COMMENTED** [review]
Fix seems to be working well - tested with android pixel 3 emulator 

left some minor suggestions

File: app/components/Views/ApproveView/Approve/index.js (hunk 4 above)
Comment (blackdevelopa) [reply app/components/Views/ApproveView/Approve/index.js]: Thank you! https://github.com/MetaMask/metamask-mobile/pull/5718/commits/0045c85ca218a3990c90080dfba9c789f12ace61

File: app/components/UI/TransactionReview/TransactionReviewInformation/index.js [hunk 5]
```diff
@@ -221,15 +222,11 @@ class TransactionReviewInformation extends PureComponent {
...
-
-  setNetworkNonce = async () => {
-    const { setNonce, setProposedNonce, transaction } = this.props;
-    const proposedNonce = await getNetworkNonce(transaction);
-    setNonce(proposedNonce);
-    setProposedNonce(proposedNonce);
+    const { showCustomNonce, setNonce, setProposedNonce, transaction } =
+      this.props;
+    if (showCustomNonce) {
+      NetworkNonce({ setNonce, setProposedNonce, transaction });
```
[the remark this answers is not in the document]
Comment (blackdevelopa) [reply app/components/UI/TransactionReview/TransactionReviewInformation/index.js]: Updated. https://github.com/MetaMask/metamask-mobile/pull/5718/commits/0045c85ca218a3990c90080dfba9c789f12ace61

File: app/components/UI/TransactionReview/TransactionReviewInformation/index.js [hunk 6]
```diff
@@ -221,15 +222,11 @@ class TransactionReviewInformation extends PureComponent {
   };
 
   componentDidMount = async () => {
```
[the remark this answers is not in the document]
Comment (blackdevelopa) [reply app/components/UI/TransactionReview/TransactionReviewInformation/index.js]: https://github.com/MetaMask/metamask-mobile/pull/5718/commits/0045c85ca218a3990c90080dfba9c789f12ace61

File: app/util/networks/networkNonce.ts [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+  setProposedNonce: (nonce: number) => void;
+  from: string;
+}
+
+const setNetworkNonce = async ({
+  setNonce,
+  setProposedNonce,
+  from,
+}: NetworkNonceProps) => {
+  const proposedNonce = await getNetworkNonce({ from });
```
Comment (digiwand) [inline app/util/networks/networkNonce.ts:14]: Is there a specific reason we want to pass the object to destruct it? 

`app/util/networks/index.js`:
```
export const getNetworkNonce = async ({ from }) => {
```

I imagine it's for consistency and wondering if there are notes somewhere for this coding practice

**Review by digiwand — APPROVED** [review]
lgtm! no blockers, just 1 question above

File: app/util/networks/networkNonce.ts (hunk 7 above)
Comment (blackdevelopa) [reply app/util/networks/networkNonce.ts:14]: Yeah, it's definitely for consistency. We can definitely choose not to destructure it here and just pass the needed `from` prop given that's all it needs.

**Comment by jpuri** [comment]
> > I am thinking if this logic can be put in 1 place rather than repeating in 3 react components.
> 
> I think it is? No? [Here](https://github.com/MetaMask/metamask-mobile/blob/95f741c56e0da0fccd7708b35f62746e316d8d85/app/util/networks/networkNonce.ts).

If we put this logic in `CustomNonce/index.js` we do not need it at these 3 places in transaction flow. Code will be clean and redundancy removed.

File: app/util/networks/networkNonce.ts (hunk 7 above)
Comment (blackdevelopa) [reply app/util/networks/networkNonce.ts]: https://github.com/MetaMask/metamask-mobile/pull/5718/commits/648c663084b1727c781447baf8736a8025e4e9ce

File: app/components/Views/ApproveView/Approve/index.js [hunk 8]
```diff
@@ -192,15 +209,23 @@ class Approve extends PureComponent {
...
+    const { showCustomNonce } = this.props;
     if (!this.props?.transaction?.id) {
       this.props.toggleApproveModal(false);
       return null;
     }
     if (!this.props?.transaction?.gas) this.handleGetGasLimit();
 
     this.startPolling();
-
+    showCustomNonce && (await this.setNetworkNonce());
```
Comment (jpuri) [inline app/components/Views/ApproveView/Approve/index.js:228]: NIT: I would prefer to write this like
```
if (showCustomNonce) {
    await this.setNetworkNonce();
}
```

**Review by jpuri — APPROVED** [review]

**Comment by seaona** [comment]
I am still able to reproduce the original issue `Submitting transaction from the same address that was just used for tx on desktop gives error, i.e. nonce doesn't get calculated on transaction prepare`.

It looks like nonce is not updated, even after waiting some minutes. It seems it only updates if we change the network and then go back to the previous network.

https://user-images.githubusercontent.com/54408225/226920168-489aa174-7b51-41c1-b618-9bcea1fbe16f.mp4

**Repro:**
1. Open a walllet on MM Extension 
2. Import the same wallet on MM Mobile
3. Select the same Account and Network on both platforms
4. Perform a Send on MM Extension -- check nonce value
5. Wait until tx is confirmed
6. Perform a Send on MM Mobile -- check nonce value - see it is not updated, and JSON RPC error appears

**Comment by blackdevelopa** [comment]
Hey @seaona, thanks for spotting this. I have now made an update for the second issue. See recording [here](http://recordit.co/E8MH5pVXcv). I am unable to reproduce the first, kindly confirm if this solves for the first.

File: app/components/Views/ApproveView/Approve/index.js [hunk 9]
```diff
@@ -374,6 +401,10 @@ class Approve extends PureComponent {
       );
     }
 
+    if (showCustomNonce && transactionToSend.nonce) {
+      transactionToSend.nonce = BNToHex(transactionToSend.nonce);
+    }
```
Comment (jpuri) [inline app/components/Views/ApproveView/Approve/index.js:406]: Hey @blackdevelopa : can you plz detail above lines, I am confused why is this conversion required.

Comment (blackdevelopa) [reply app/components/Views/ApproveView/Approve/index.js:406]: Hey @jpuri, sure. The transaction object expects the nonce value to be converted to hex and we do this at the point of creating the txn object. We also see that here when rendering the transactions in the ActivityView [here](https://github.com/MetaMask/metamask-mobile/blob/edfe239b20c018dc46cae39765353b3fe1825c18/app/components/UI/TransactionElement/TransactionDetails/index.js#L358)

**Review by digiwand — APPROVED** [review]

**Comment by seaona** [comment]
The first point was just a confirmation that the main flow was working good, so not actually an issue. :smiley: 
I've re-tested again and I see the issue fixed. Overll looks good @blackdevelopa. The PR can be merged

**Pull request merged**